modules/workflow-service-api/src/main/java/org/opencastproject/workflow/api/WorkflowIdentifier.java

```diff
@@ -0,0 +1,93 @@
+/**
+ * Licensed to The Apereo Foundation under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ *
+ * The Apereo Foundation licenses this file to you under the Educational
+ * Community License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License
+ * at:
+ *
+ *   http://opensource.org/licenses/ecl2.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package org.opencastproject.workflow.api;
+
+import java.util.Objects;
+
+/**
+ * Represents a workflow's "primary key", for use in maps and sets (immutable)
+ */
+public final class WorkflowIdentifier {
+  private final String id;
+  private final String organization;
+
+  /**
+   * Create an identifier directly from a workflow definition
+   * @param definition the workflow definition
+   * @return the constructed identifier
+   */
+  public static WorkflowIdentifier fromDefinition(final WorkflowDefinition definition) {
```

Comment:
Why aren't these simply constructors?

Reply:
It's simply a matter of style, of course, but I find multiple overloaded constructors

1. hard to read
2. don't scale

Let me illustrate the first point:

```
new WorkflowDefinition("foo", null)
```

Reading this, do you know what the `null` is for?

As for the second point, there are lots of examples in Opencast, but you might take a look at `SchedulerItem`. As you can see, not only has this class many constructors, but they partly overlap. One takes a `String`, one takes `String` and a `long`, which isn't even used. This is done (generally speaking) so that the compiler can figure out which constructor you mean.

In order to combat this, I invent this named, static "constructors".

Reply:
Okay, I get what you're saying, but I still think this is relatively straight-forward in this case: One constructor that takes both id and organization, one that only takes the id, one that takes a workflow definition. So no need for WorkflowIdentifier("id", null), except internally. These are all easily distinguishable by the compiler.